Minimal implementation of branch coverage in LCOV parser

magneland opened the pull request:
A minor translation from '-' to 0 avoids extra type checking later.
Also deletes a couple of unused methods.
TODO: fix all the rubocop violations in `parse_line`

Branch coverage information is stored which one line per branch:
```
  BRDA:<line number>,<block number>,<branch number>,<taken>

  Block  number  and  branch  number are gcc internal IDs for the branch.
  Taken is either '-' if the basic block containing the branch was  never
  executed or a number indicating how often that branch was taken.
```
See: http://ltp.sourceforge.net/coverage/lcov/geninfo.1.php

magneland commented:
This implements #26 by using the following heuristic:

If using branch coverage, any branch not taken in a line marks the line as uncovered.

grodowski commented:
@magneland thanks for your contribution to undercover, I'm so excited to give it a closer look and a test drive (hopefully very soon) 🙌

grodowski reviewed (commented):
> If using branch coverage, any branch not taken in a line marks the line as uncovered.

This is a great starting point with more precise output and reporting to be added later.

Left some comments and opened https://github.com/magneland/undercover/pull/1 to work on an integration spec in `report_spec.rb`. I think we should be able to merge it very soon!

magneland commented on lib/undercover/lcov_parser.rb:
```diff
@@ -41,12 +41,22 @@ def parse_line(line)
+      when /^BRF:(\d+)/
+        # TODO: remember branches found
+      when /^BRH:(\d+)/
+        # TODO: remember branches hit
```
I think in the interest of simplicity it makes sense to remove these for now.

magneland replied:
Actually I realized there's a catch-all `else` below, so it needs to be handled somehow.
I tweaked this chunk to be more understandable, hopefully.

magneland commented on lib/undercover/lcov_parser.rb:
```diff
@@ -41,12 +41,22 @@ def parse_line(line)
+        # TODO: remember branches hit
+      when /^BRDA:(\d+),(\d+),(\d+),(-|\d+)/
+        line_no = $~[1]
+        block_no = $~[2]
```
That's a good catch. I believe that could be considered a bug in `simplecov-lcov`.
This block number concept comes from LCOV and not "Ruby blocks".
However, I noticed that @mame, the implementer of the original `coverage.so` library, is making use of it as a secondary counter here: https://github.com/mame/coverage-helpers/blob/4d2a33d3dec8c8bbf51bb21a3cdaf57d2481472e/lib/coverage/helpers.rb#L263

grodowski replied:
Ok, according to the implementation you shared this is numbering consecutive "branchable" expressions in each reported file. Found a [test](https://github.com/mame/coverage-helpers/blob/4d2a33d3dec8c8bbf51bb21a3cdaf57d2481472e/test/coverage/helpers_test.rb#L378)  for `to_lcov_info` - it doesn't make any expectations on this particular value. Anyhow, it's probably fine to keep `parse_line` like this?

magneland replied:
Sounds good. I think it probably doesn't matter for `undercover` but maybe `simplecov-lcov` could be changed.

magneland commented on lib/undercover/result.rb:
```diff
@@ -27,18 +27,11 @@ def flagged?
-    end
-
     def uncovered?(line_no)
+      coverage.each do |ln, _block, _branch, cov|
+        return true if ln == line_no && cov.is_a?(Numeric) && cov.zero?
+      end
```
Yeah I was kind of on the fence about that. I agree it's not necessary.

magneland replied:
Although - I guess the real reason for that check was that `cov` can be `nil` (actually it most cases it's `nil`).
I will replace it with a `nil` check.

magneland commented on undercover.gemspec:
```diff
@@ -34,7 +34,7 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency 'simplecov'
   spec.add_development_dependency 'simplecov-html'
-  spec.add_development_dependency 'simplecov-lcov'
+  spec.add_development_dependency 'simplecov-lcov', '~> 0.8' # branch coverage
```
Fixing. I was meaning to remove that :-)

magneland commented on spec/lcov_parser_spec.rb:
```diff
@@ -18,6 +18,18 @@
   end
 
+  it 'reports a structure of source files with branch coverage' do
+    parser = described_class.parse('spec/fixtures/imagen_rb.lcov')
```
Yep I agree. I decided to reuse `imagen_rb` since you already used it (and wrote it).
But I would say the ideal would be to use some exemplary code as input instead.

grodowski replied:
Can you remove this files and create a new fixture file or update `spec/fixtures/sample.lcov` that's used in the spec example above?

magneland commented:
Thanks for the review @grodowski :-) 
I have added 3 fixup commits addressing your comments.
Let me know if you think it's good enough now, or if more tweaking is needed.
Also good point with the integration spec. I forgot to look at that.

grodowski commented on spec/report_spec.rb:
```diff
@@ -34,14 +34,19 @@
 
+    # includes flagged blocks with incomplete branch coverage
+    expect(module_flagged[1].node.name).to eq('branch_missed')
+    pending 'coverage_f need to handle branch coverage data now present in Result'
```
The `pending` should be removed before we merged - right now the `coverage_f` for `branch_missed` fixture is > 0 (L48)

magneland replied:
OK that's another good catch. I made a WIP commit that refactors `coverage_f`. I still need to write some specs for that new algorithm but I don't fully understand the previous implementation or even the feature yet. So probably I will get to that next weekend...

grodowski replied:
Cool, I'd be happy to provide any help if you need it. `coverage_f` returns the "classic" line coverage percentage for a given `Result` (representing a method, class, ruby block).

magneland replied:
Thanks, I understand it now.
I was a bit confused by the existing implementation, but I believe my refactor should be correct now.

grodowski reviewed (commented):
@magneland thanks for making these fixups. I commented on two more things we should complete before merge.

Readme docs and release can be done separately. I think it's enough for the docs to mention how branch coverage works in undercover and that [branch coverage](https://github.com/simplecov-ruby/simplecov#branch-coverage-ruby--25) needs to be enabled.

magneland commented:
I believe this is in a pretty good state now. Let me know at your leisure what you think!

BTW, one I thing I didn't quite get is how realistic `fixtures.lcov` is meant to be. Is it supposed to be the output of `class.rb` and `module.rb`? If so, I am confused on how the lines corresponding to `end` statements got there.

grodowski commented:
@magneland yeah the `fixture.lcov` file is handcrafted as of now and I agree that it's confusing. A potential improvement could be to create some actual fixture Ruby files with corresponding specs that could be used anytime to regenerate a correct `fixture.lcov`

grodowski reviewed (approved):
Thanks for your contribution @magneland!

grodowski commented:
👀 https://github.com/grodowski/undercover/commit/debdaba0f05f91f3bb0894803f1b0c8612cd66c5. I'll be documenting the feature in `README` and releasing a new minor version soon
